### DISCUSSION TITLE
Set SharedStringTable (sst)'s count and uniqueCount

### DISCUSSION THREAD

**Discussion by ogaken-1** [ClosedXML/ClosedXML#2023] [Ideas]
I want to change SharedStringTableWriter's behavior to set `count` and `uniqueCount` attributes,
because they are specified to not optional in MS-OI29500.

Cited below.

```
2.1.676 Part 1 Secition 18.4.9, sst (Shared String Table)

a. The standard specifies uniqueCount as an unbound unsigned int.
   Office requires the value be less than or equal to 2147483647.

b. The standard specifies count as an unbound unsigned int.
   Office requires the value be less than or equal to 2147483647.
```

I think that the process of writing using XmlWriter and make the sst table were happening at the same time seemed to make the problem more difficult.

If the process of make the sst table is done in advance, it would be possible to identify the values of the count and uniqueCount attributes.

It seems to me that using XElement to create the sst table in advance would solve this problem, what do you think of this idea?

**Comment by ogaken-1** [comment]
I made a modification in the fork repository to use XElement, and it seems to be working.

However, I am hesitant to send this to you due to performance concerns and the fact that I cannot run the tests in my local environment (Linux). (In fact, this doesn't seem like a good patch - I'm not sure if it's a good patch or not).

Not too keen on setting sst count and uniqueCount?

**Comment by jahav** [comment]
> Not too keen on setting sst count and uniqueCount?

Not really, I would either need to enumerate cells twice or materialize it to table. It's not that few tens of MB (as an example [20 cols 100k rows] cell references + id = 12 bytes per record) would kill me, I just don't see a compelling reason.

Both attributes are optional (MS-OI29500 only adds additonal info how Excel implemeted features, in this case the allowed range of values was decreased from 4bln to 2bln, value itself is still optional). Excel can read files without attributes and developers can just use List or something and push read elements at the end.

In the long run, I would add it through #1969 (or equivalent).

**Reply by ogaken-1** [reply]
Thank you for replying.

The reason I need this is because the XLSX -> PDF converter I am trying to use works on the assumption that the count and uniqueCount attributes are written.

When I contacted the developer of the tool, I got the response that Excel writes them, so I wondered if I could solve this by changing ClosedXML.

It appears that I have misread the MS-OI regarding the content of it. Thank you for pointing this out to me.
